# MINST 필기체 숫자 인식

In [1]:
import tensorflow as tf

In [2]:
import numpy as np

In [3]:
from tensorflow import keras

In [4]:
# 신경망과 훈련 매개변수
epochs = 200
batch_size = 128
verbose = 1
nb_classes = 10     # 출력 개수 = 숫자의 개수
n_hidden = 128
validation_split = 0.2     # 검증을 위해 남겨둔 훈련 데이터

In [5]:
# MINST 데이터셋 로드
mnist = keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()

In [6]:
x_train.shape

(60000, 28, 28)

In [7]:
# 이를 28x28에서 60000, 784로 바꿔줌
reshaped = 784
x_train = x_train.reshape(60000, reshaped)
x_test = x_test.reshape(10000, reshaped)

In [8]:
x_train.shape

(60000, 784)

In [9]:
x_train = x_train.astype('float32')
x_test = x_test.astype('float32')

In [10]:
# 픽셀값은 0~255인데 이를 255로 나눠줌으로써 0~1 사이로 정규화시킴
x_train /= 255
x_test /= 255
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

60000 train samples
10000 test samples


In [11]:
# 원핫 인코딩
y_train = tf.keras.utils.to_categorical(y_train, nb_classes)
y_test = tf.keras.utils.to_categorical(y_test, nb_classes)

In [12]:
dropout = 0.3

In [13]:
# 모델 구축
model = tf.keras.models.Sequential()
model.add(keras.layers.Dense(n_hidden,
                            input_shape = (reshaped,),
                            name = 'dense_layer',
                            activation = 'relu'))
model.add(keras.layers.Dropout(dropout))
model.add(keras.layers.Dense(n_hidden,
                            name = 'dense_layer_2',
                            activation = 'relu'))
model.add(keras.layers.Dropout(dropout))
model.add(keras.layers.Dense(nb_classes,
                            name = 'dense_layer_3',
                            activation = 'softmax'))
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_layer (Dense)          (None, 128)               100480    
_________________________________________________________________
dropout (Dropout)            (None, 128)               0         
_________________________________________________________________
dense_layer_2 (Dense)        (None, 128)               16512     
_________________________________________________________________
dropout_1 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_layer_3 (Dense)        (None, 10)                1290      
Total params: 118,282
Trainable params: 118,282
Non-trainable params: 0
_________________________________________________________________


## 똑같이 따와서 이부분만 달라짐

In [15]:
# 모델 컴파일
model.compile(optimizer = 'RMSProp',   # 확률적 그래디언트 하강
             loss = 'categorical_crossentropy',
             metrics = ['accuracy'])

In [16]:
# 모델 훈련
model.fit(x_train, y_train, 
          batch_size = batch_size, 
          epochs = epochs, 
         verbose = verbose, 
         validation_split =  validation_split)

Train on 48000 samples, validate on 12000 samples
Epoch 1/200
48000/48000 [==============================] - 1s 26us/sample - loss: 0.4786 - accuracy: 0.8530 - val_loss: 0.1813 - val_accuracy: 0.9469
Epoch 2/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.2201 - accuracy: 0.9341 - val_loss: 0.1413 - val_accuracy: 0.9580
Epoch 3/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.1736 - accuracy: 0.9486 - val_loss: 0.1155 - val_accuracy: 0.9662
Epoch 4/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.1466 - accuracy: 0.9565 - val_loss: 0.1137 - val_accuracy: 0.9689
Epoch 5/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.1325 - accuracy: 0.9609 - val_loss: 0.1061 - val_accuracy: 0.9712
Epoch 6/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.1188 - accuracy: 0.9643 - val_loss: 0.0989 - val_accuracy: 0.9735
Epoch 7/200
48000/48000 [=========================

Epoch 55/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.0516 - accuracy: 0.9870 - val_loss: 0.1653 - val_accuracy: 0.9786
Epoch 56/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.0560 - accuracy: 0.9858 - val_loss: 0.1668 - val_accuracy: 0.9787
Epoch 57/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.0515 - accuracy: 0.9864 - val_loss: 0.1697 - val_accuracy: 0.9787
Epoch 58/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.0523 - accuracy: 0.9869 - val_loss: 0.1682 - val_accuracy: 0.9789
Epoch 59/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.0505 - accuracy: 0.9869 - val_loss: 0.1649 - val_accuracy: 0.9791
Epoch 60/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.0507 - accuracy: 0.9874 - val_loss: 0.1770 - val_accuracy: 0.9804
Epoch 61/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.0533 - ac

48000/48000 [==============================] - 1s 14us/sample - loss: 0.0524 - accuracy: 0.9883 - val_loss: 0.2452 - val_accuracy: 0.9780
Epoch 110/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.0520 - accuracy: 0.9888 - val_loss: 0.2452 - val_accuracy: 0.9781
Epoch 111/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.0501 - accuracy: 0.9887 - val_loss: 0.2394 - val_accuracy: 0.9787
Epoch 112/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.0486 - accuracy: 0.9884 - val_loss: 0.2470 - val_accuracy: 0.9778
Epoch 113/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.0526 - accuracy: 0.9890 - val_loss: 0.2205 - val_accuracy: 0.9786
Epoch 114/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.0466 - accuracy: 0.9887 - val_loss: 0.2299 - val_accuracy: 0.9793
Epoch 115/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.0525 - accuracy:

Epoch 163/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.0475 - accuracy: 0.9896 - val_loss: 0.2908 - val_accuracy: 0.9786
Epoch 164/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.0514 - accuracy: 0.9901 - val_loss: 0.3164 - val_accuracy: 0.9772
Epoch 165/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.0508 - accuracy: 0.9898 - val_loss: 0.3101 - val_accuracy: 0.9786
Epoch 166/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.0507 - accuracy: 0.9899 - val_loss: 0.2909 - val_accuracy: 0.9780
Epoch 167/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.0435 - accuracy: 0.9900 - val_loss: 0.2890 - val_accuracy: 0.9793
Epoch 168/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.0556 - accuracy: 0.9885 - val_loss: 0.2816 - val_accuracy: 0.9788
Epoch 169/200
48000/48000 [==============================] - 1s 14us/sample - loss: 0.04

In [17]:
# 모델 평가
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy : ',test_acc)

10000/10000 [==============================] - 0s 18us/sample - loss: 0.3098 - accuracy: 0.9773
Test accuracy :  0.9773


### 정확도 97%
### SGD에 비해 보다 빠르게 최적화되어 정확도를 얻을 수 있음